File: Leetcode/0034 Find First and Last Position of Element in Sorted Array.cpp
class Solution {
public:
    enum SerachType { FIRST, LAST };

    vector<int> searchRange(vector<int>& nums, int target) {
        return {binarySearch(nums,target,FIRST),binarySearch(nums,target,LAST)};        
    }

    int binarySearch(vector<int> a, int val, SerachType type) {
        int lo = 0;
        int hi = (int)a.size() - 1;
        int mid;
        int index = -1;
        while (lo <= hi) {
            mid = (lo + hi) / 2;
            if (a[mid] == val) {
                index = mid;
                if (type == FIRST)
                    hi = mid - 1;
                if (type == LAST)
                    lo = mid + 1;
            } else if (a[mid] < val)
                lo = mid + 1;
            else
                hi = mid - 1;
        }
        return index;
    }
};

File: Leetcode/0034 Find First and Last Position of Element in Sorted Array.java
class Solution {
    public enum OccurrenceType {
        ANY, FIRST, LAST
    }

    public int binarySearch(int[] sortedNumbers, int target, OccurrenceType occurrenceType) {
        int lo = 0, hi = sortedNumbers.length - 1, occurrenceIndex = -1;
        while (lo <= hi) {
            int mid = lo + (hi - lo) / 2;
            if (sortedNumbers[mid] == target) {
                if (occurrenceType == OccurrenceType.ANY) {
                    return mid;
                }
                occurrenceIndex = mid;
                if (occurrenceType == OccurrenceType.FIRST) {
                    hi = mid - 1;
                } else {
                    lo = mid + 1;
                }
            } else if (sortedNumbers[mid] > target) {
                hi = mid - 1;
            } else {
                lo = mid + 1;
            }
        }
        return occurrenceIndex;
    }
    
    public int[] searchRange(int[] sortedArray, int target) {
        int firstIndex = binarySearch(sortedArray, target, OccurrenceType.FIRST);
        int lastIndex = binarySearch(sortedArray, target, OccurrenceType.LAST);
        return new int[]{firstIndex, lastIndex};
    }
}
